HPCC-16543 Improve error reporting of JlibSemTest and implement JlibSemTestStress.

AttilaVamos opened the pull request:
Add time and taken values to ASSERT log

Add semaphore stress test

Signed-off-by: Attila Vamos attila.vamos@gmail.com

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3c12047bf40c6d39cf4d6da0eed15faa4de476e2
Build: success
Install hpccsystems-platform-community_6.3.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
  a Test a 
Sha: 3c12047bf40c6d39cf4d6da0eed15faa4de476e2
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.17 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest
Sha: 3c12047bf40c6d39cf4d6da0eed15faa4de476e2
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.625 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -57,12 +57,15 @@ class JlibSemTest : public CppUnit::TestFixture
         sem.wait(time);
         unsigned taken = msTick() - now;
-        ASSERT(taken >= time && taken < 2*time);
+        VStringBuffer errMsg("values: time: %d, taken: %d", time, taken);
```
trivial: %u more correct.

AttilaVamos replied:
Ok

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
+
+    void testCleanup()
+    {
+    }
```
Not new, but good time to clearup/remove empty test.

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
+
+    void testSetup()
+    {
+    }
```
Not new, but good time to clearup/remove empty test.

AttilaVamos replied:
Removed

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
+        for (int i = 0 ; i <= loopCount; i++)
+        {
+            unsigned now = msTick();
+            sem.wait(time);
```
pedantic: could test if signalled, which would be unexpected/error.

AttilaVamos replied:
Implemented

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
+
+    void testTimedElapsed(Semaphore & sem, unsigned time, unsigned loopCount)
+    {
+        unsigned long sumTaken = 0;
```
minor: if you want a 8 byte integer, we use a unsigned __int64, which guarantees it,

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
+        //Very basic semaphore stress tests.
+        Semaphore sem;
+        sem.signal();
+        sem.wait();
```
minor: should test that actually signalled: (if (sem.wait())

AttilaVamos replied:
Implemented

jakesmith replied:
Right, but probably should have timeout here e.g. if (!sem.wait(1000)) error, better than stalling.
Should never have to wait of course.

jakesmith commented:
@AttilaVamos - a few trivial/minor comments

AttilaVamos commented:
@jakesmith updated, please check again.

HPCCSmoketest commented:
Automated Smoketest
Sha: 7e834bde1ff48e0851c870eb64276f55990d3265
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.149 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -93,42 +83,37 @@ class JlibSemTestStress : public CppUnit::TestFixture
-        CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0 == numberOfOut);
+        VStringBuffer errMsg("values: time: %d, loop: %d, sum taken: %llu, average taken: %llu, max taken: %d, out of limit: %d times, signaled: %s",
+                                time, loopCount, sumTaken, sumTaken/loopCount, maxTaken, numberOfOut, (isSignaled ? "yes" : "no"));
+        CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0 == numberOfOut || isSignaled );
```
This condition look wrong.
Should be : 0 == numberOfOut && !isSignaled

jakesmith commented:
@AttilaVamos - one issue. If you want to go ahead and squash after changing that, then will be ready to merge.

AttilaVamos committed:
HPCC-16543 Improve error reporting of JlibSemTest and implement
JlibSemTestStress.

Add time and taken values to ASSERT log

Add semaphore stress test

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith updated and squashed, please take a look on it.

HPCCSmoketest commented:
Automated Smoketest
Sha: 19f0f24ac77c743d7c4d274d9b88b66be12c8cb4
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 89.361 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -85,6 +78,62 @@ class JlibSemTest : public CppUnit::TestFixture
+        if (!sem.wait(1000))
+        {
+            VStringBuffer errMsg("Semaphore stalled (%s:%s)", __FILE__, __LINE__);
+            CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0);
```
trivial: Could you have used CPPUNIT_FAIL, since you have no condition here?

jakesmith commented:
@richardkchapman - good to merge.

AttilaVamos commented:
@richardkchapman please wait with merge until I will change to CPPUNIT_FAIL().

HPCCSmoketest commented:
Automated Smoketest
Sha: 5ba27bde5c962bc9faecb46234e368c68e2cc977
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.653 seconds |

Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|86 |86 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@richardkchapman please merge.
